src/main/java/htsjdk/samtools/SamReaderFactory.java

```diff
@@ -311,7 +311,24 @@ public SamReader open(final SamInputResource resource) {
                     if (SamStreams.isBAMFile(bufferedStream)) {
                         if (sourceFile == null || !sourceFile.isFile()) {
                             // Handle case in which file is a named pipe, e.g. /dev/stdin or created by mkfifo
-                            primitiveSamReader = new BAMFileReader(bufferedStream, indexFile, false, asynchronousIO, validationStringency, this.samRecordFactory);
+                            if (indexFile!=null) {
+                                primitiveSamReader = new BAMFileReader(bufferedStream, indexFile, false, asynchronousIO, validationStringency, this.samRecordFactory);
+                            } else {
+                                final SeekableStream indexSeekable = indexMaybe == null ? null : indexMaybe.asUnbufferedSeekableStream();
+                                // do not close bufferedStream, it's the same stream we're getting here.
+                                SeekableStream sourceSeekable = data.asUnbufferedSeekableStream();
+                                if (null != sourceSeekable && null != indexSeekable) {
+                                    // let's make sure the user can seek in this file.
+                                    // need to return to the beginning because it's the same stream we used earlier
+                                    // and read a bit from.
+                                    sourceSeekable.seek(0);
+                                    primitiveSamReader = new BAMFileReader(
+                                        sourceSeekable, indexSeekable, false, asynchronousIO, validationStringency, this.samRecordFactory);
+                                } else {
+                                    primitiveSamReader = new BAMFileReader(
```

Comment:
This else clause is the same as the `if (indexFile != null)` clause above -- is it possible to simplify this nested if/else to combine them into a single case?